### PR TITLE
[NavigationBar] - Update handling of center-aligned titles

### PR DESCRIPTION
**Pull request by ShepJGoogle**
Update handling of center-aligned titles to follow Apple’s implementation of shifting title over if leading buttons encroach.

Rule -

If MDCNavigationBarTitleAlignmentCenter:
- If leading button bar's maximum x coordinate is less than centered titleFrame's x origin, no change (title stays center aligned)
- else, use leading title alignment

This will close Issue #1457 

Apple's UINavigation Bar Alignment Rules:
screen shot 2017-06-02 at 3 32 25 pm

screen shot 2017-06-02 at 3 32 57 pm

screen shot 2017-06-02 at 3 33 21 pm

screen shot 2017-06-02 at 3 33 42 pm

After Change:
screen shot 2017-06-02 at 3 27 31 pm

screen shot 2017-06-02 at 3 28 01 pm

screen shot 2017-06-02 at 3 28 23 pm

screen shot 2017-06-02 at 3 29 06 pm

### CODE REVIEW

File: components/NavigationBar/src/MDCNavigationBar.m [hunk 1]
```diff
@@ -460,10 +460,13 @@ - (CGRect)mdc_frameAlignedVertically:(CGRect)frame
...
-      return CGRectMake(CGRectGetMaxX(self.bounds) / 2 - frame.size.width / 2, frame.origin.y,
-                        frame.size.width, frame.size.height);
-
+    case MDCNavigationBarTitleAlignmentCenter: {
+      CGFloat xOrigin = CGRectGetMaxX(self.bounds) / 2 - frame.size.width / 2;
+      if (frame.origin.x <= xOrigin) {
+        return CGRectMake(xOrigin, frame.origin.y,
+                          frame.size.width, frame.size.height);
+      }
+    }
```
Comment (ajsecord) [inline components/NavigationBar/src/MDCNavigationBar.m]: Are you deliberately falling through to the next case here? Intentional falling through [needs a comment](https://engdoc.corp.google.com/eng/doc/objcguide.xml?external=y&showone=Conditionals#Conditionals).

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: components/NavigationBar/src/MDCNavigationBar.m (hunk 1 above)
Comment (ShepJGoogle) [reply components/NavigationBar/src/MDCNavigationBar.m]: Done

File: components/NavigationBar/src/MDCNavigationBar.m [hunk 2]
```diff
@@ -460,10 +460,14 @@ - (CGRect)mdc_frameAlignedVertically:(CGRect)frame
 - (CGRect)mdc_frameAlignedHorizontally:(CGRect)frame
                              alignment:(MDCNavigationBarTitleAlignment)alignment {
   switch (alignment) {
-    case MDCNavigationBarTitleAlignmentCenter:
-      return CGRectMake(CGRectGetMaxX(self.bounds) / 2 - frame.size.width / 2, frame.origin.y,
-                        frame.size.width, frame.size.height);
-
+    case MDCNavigationBarTitleAlignmentCenter: {
```
Comment (ajsecord) [inline components/NavigationBar/src/MDCNavigationBar.m]: Add comment like "Align title to the center, except if X, Y, or Z." This isn't clear enough to just parse.

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: components/NavigationBar/src/MDCNavigationBar.m (hunk 2 above)
Comment (ShepJGoogle) [reply components/NavigationBar/src/MDCNavigationBar.m]: Done

File: components/NavigationBar/src/MDCNavigationBar.m [hunk 3]
```diff
@@ -460,10 +460,15 @@ - (CGRect)mdc_frameAlignedVertically:(CGRect)frame
 - (CGRect)mdc_frameAlignedHorizontally:(CGRect)frame
                              alignment:(MDCNavigationBarTitleAlignment)alignment {
   switch (alignment) {
-    case MDCNavigationBarTitleAlignmentCenter:
-      return CGRectMake(CGRectGetMaxX(self.bounds) / 2 - frame.size.width / 2, frame.origin.y,
-                        frame.size.width, frame.size.height);
-
+      // Center align title if desired unless leading icons will overlap frame
+    case MDCNavigationBarTitleAlignmentCenter: {
+      CGFloat xOrigin = CGRectGetMaxX(self.bounds) / 2 - frame.size.width / 2;
```
Comment (ianegordon) [inline components/NavigationBar/src/MDCNavigationBar.m]: Avoid frame.size or frame.origin unless you are positive the frame has been standardized.  (And in that case, document that it is).
s/frame.size.width/CGRectGetWidth(frame)/
(for height as well)

**Review by ianegordon — CHANGES_REQUESTED** [review]

**Comment by ianegordon** [comment]
Strange that clang didn't through a warning on the switch fallthrough :/

**Review by ajsecord — APPROVED** [review]

**Comment by ianegordon** [comment]
rect.origin -> CGRectGetMin()

Consider simplifying the next line to just return frame.

**Review by ajsecord — APPROVED** [review]

**Pull request merged**